src/Planner/findParallelReplicasQuery.cpp

```diff
@@ -55,17 +76,20 @@ std::vector<const QueryNode *> getSupportingParallelReplicasQuery(const IQueryTr
         {
             case QueryTreeNodeType::TABLE:
             {
-                const auto & table_node = query_tree_node->as<TableNode &>();
-                const auto & storage = table_node.getStorage();
-                /// Here we check StorageDummy as well, to support a query tree with replaced storages.
-                if (std::dynamic_pointer_cast<MergeTreeData>(storage) || typeid_cast<const StorageDummy *>(storage.get()))
+                /// Storage may have been replaced with StorageDummy for analysis, but here we need
```

Comment:
It can be simplified by setting necessary information for StorageDummy(). This way, we can simply check if it's `isMergeTree()`, `supportsReplication()`. The same for dummy table node, - here we just need to store original modifiers.

I've just tried this approach in https://github.com/ClickHouse/ClickHouse/pull/73584, but something is off there, - tests crashing